cpp/src/arrow/util/io_util.cc

```diff
@@ -1734,5 +1665,21 @@ int64_t GetRandomSeed() {
   return static_cast<int64_t>(seed_gen());
 }
 
+uint64_t GetThreadId() {
+  uint64_t equiv{0};
+  // std::thread::id is trivially copyable as per C++ spec,
+  // so type punning as a uint64_t should work
+  static_assert(sizeof(std::thread::id) <= sizeof(uint64_t),
+                "std::thread::id can't fit into uint64_t");
+  const auto tid = std::this_thread::get_id();
+  memcpy(&equiv, reinterpret_cast<const void*>(&tid), sizeof(tid));
+  return equiv;
+}
+
+uint64_t GetOptionalThreadId() {
+  auto tid = GetThreadId();
+  return (tid == 0) ? tid - 1 : tid;
```

Comment:
If we're going to use -1 in multiple places like this, let's define it as a global constant:
```suggestion
  return (tid == 0) ? kUnlikelyThreadId : tid;
```

Reply:
Actually I'm not sure I understand the intent of this function and it's not exported

Reply:
Hmm, that's a remnant from a previous approach, sorry. I should simply remove it.